src/docker/composeappengine.cc

```diff
@@ -41,72 +41,113 @@ AppEngine::Result ComposeAppEngine::fetch(const App& app) {
   return result;
 }
 
-bool ComposeAppEngine::install(const App& app) {
-  bool result = false;
+AppEngine::Result ComposeAppEngine::install(const App& app) {
+  Result result{false};
 
+  // check if App dir exists
   if (!boost::filesystem::exists(appRoot(app))) {
-    LOG_ERROR << "App dir doesn't exist, cannot install App that hasn't been fetched";
-    return result;
+    return {false, "App dir doesn't exist, cannot install App that hasn't been fetched"};
   }
 
-  try {
-    AppState state(app, appRoot(app));
-
-    switch (state()) {
-      case AppState::State::kPulled:
-      case AppState::State::kInstallFail: {
+  AppState state(app, appRoot(app));
+
+  // do App install if not installed yet
+  switch (state()) {
+    case AppState::State::kInstalled:
+    case AppState::State::kStarted:
+      LOG_DEBUG << app.name << " has been already installed/started";
+      result = true;
+      break;
+    case AppState::State::kPulled: {
+      try {
         installApp(app);
-        if (!areContainersCreated(app)) {
-          state.setState(AppState::State::kInstallFail);
-          break;
-        }
-        state.setState(AppState::State::kInstalled);
-      }
-      case AppState::State::kInstalled:
-      case AppState::State::kStarted:
         result = true;
-        break;
-      default:
-        LOG_ERROR << "Cannot install App that hasn't been fetched";
+      } catch (const std::exception& exc) {
+        result = {false, exc.what()};
+      }
+      break;
+    }
+    default:
+      result = {false, "Cannot install App that hasn't been fetched"};
+      break;
+  }
+
+  if (!result) {
+    return result;
+  }
+
+  // check whether App's containers have been actually created, if possible
+  try {
+    if (!areContainersCreated(app)) {
+      result = {false, "App containers haven't been created"};
     }
   } catch (const std::exception& exc) {
-    LOG_WARNING << "Failed to get/set App state, fallback to forced install: " << exc.what();
+    LOG_WARNING << "Failed to check whether containers have been created: " << exc.what();
+    // if we fail (exception) to check whether containers have been created is doesn't mean
+    // that installation was not succesful, just assume that the App creation succeeded if
+    // `docker-compose up` returns EXIT_SUCCESS
```

Comment:
would this be something like a one-shot container already exiting?

Reply:
Could be, a container, in this case, still be present, but at "exited" state.
This is rather for the case when aklite fails to inquire/request a list of containers from dockerd because of some unknown reason (exception), e.g. no `services:` in docker-compose.yml, or dockerd is not running at this moment or not listening socket.